### DISCUSSION TITLE
CORS policy : preflight issue

### DISCUSSION THREAD

**Discussion by guigui42** [zitadel/zitadel#8951] [Q&A] [answered]
my app uses golang and zitadel-go sdk.

when I push it to fly.io, and open some pages, I get some CORS errors in the console :  
```Access to XMLHttpRequest at 'https://xxxxxx-zrzu8s.us1.zitadel.cloud/oauth/v2/authorize?client_id=294068385057176497&code_challenge=0TIns3Fpo4GYFpxOYhHLI0EpaR9W87T7opsTwKYV0Ic&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fxxxxxx.fly.dev%2Fauth%2Fcallback&response_type=code&scope=openid+profile+email+urn%3Azitadel%3Aiam%3Auser%3Aresourceowner+urn%3Azitadel%3Aiam%3Auser%3Ametadata&state=QhZvFvT_uk7P8b_45rpYWi2Vnc_KcBygOBX3t3WboPOVFI66kaIOCQx3Hr9aJw' (redirected from 'https://xxxxxx.fly.dev/api/images') from origin 'https://xxxxxx.fly.dev' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.```

looking at the request : 

```
Request URL:
https://xxxxxx-zrzu8s.us1.zitadel.cloud/oauth/v2/authorize?client_id=294068385057176497&code_challenge=Xf7_1wctaebIrIyggKcJM_1mxaAxaEBieoWyMjqoLG8&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fxxxxxx.fly.dev%2Fauth%2Fcallback&response_type=code&scope=openid+profile+email+urn%3Azitadel%3Aiam%3Auser%3Aresourceowner+urn%3Azitadel%3Aiam%3Auser%3Ametadata&state=c2zoSVCDOH72KqHkFrk7eUS1ILJvE6zbpN4Fn51PjuTmqpoKzGwCnnwIWuNcZw_JvQ

Request Method:OPTIONS
Status Code:204 No Content
Referrer Policy:strict-origin-when-cross-origin
access-control-request-headers:hx-current-url,hx-request,hx-trigger-name,x-workspace-id
access-control-request-method:GET
origin:https://2Fxxxxxx.fly.dev
referer:https://2Fxxxxxx.fly.dev/
sec-fetch-mode:cors
sec-fetch-site:cross-site
```

And the response header : 
```
server:Google Frontend
strict-transport-security:max-age=63072000; includeSubDomains; preload
traceparent:00-56940e1f34638923ec0764bf657fae90-c258b5f0d7594211-00
vary:Origin, Access-Control-Request-Method, Access-Control-Request-Headers
via:1.1 google
x-cache-hit:uncacheable
x-cloud-trace-context:56940e1f34638923ec0764bf657fae90/14004143087318286865
x-robots-tag:none
```

In my code, I setup cors as a middleware

```
	r.Use(cors.Handler(cors.Options{
		AllowedOrigins: []string{"*"}, // Use this to allow specific origin hosts
		// AllowedOrigins: []string{"https://" + conf.Conf.Domain, "https://" + conf.Conf.ZitadelIssuer},
		AllowedMethods:   []string{"GET", "POST", "PUT", "DELETE", "OPTIONS"},
		AllowedHeaders:   []string{"Accept", "Authorization", "Content-Type", "X-CSRF-Token", "hx-current-url", "hx-request", "HX-Request", "HX-Boosted", "HX-Current-URL", "HX-History-Restore-Request", "HX-Prompt", "HX-Target", "HX-Trigger-Name", "HX-Trigger", "HX-Location", "HX-Push-Url", "HX-Redirect", "HX-Refresh", "HX-Replace-Url", "HX-Reswap", "HX-Retarget", "HX-Reselect", "HX-Trigger-After-Settle", "HX-Trigger-After-Swap"},
		ExposedHeaders:   []string{"Link"},
		AllowCredentials: true,
		MaxAge:           300, // Maximum value not ignored by any of major browsers
	}))
```

Here is my configuration in Zitadel : 
![image](https://github.com/user-attachments/assets/311022c3-1c01-49dc-bb46-7420f0cc0070)
![image](https://github.com/user-attachments/assets/c2e03e40-e266-458f-b2cd-b9bd582a4a17)

And this issue does not happen when I work locally, so I am not sure if this is a fly.io or zitadel issue.

I have tried different cors configuration, * or specific, but same problem.
Could it be something else I need to configure on Zitadel side ?

**Comment by guigui42** [comment]
Also maybe a related issue : https://github.com/zitadel/zitadel/discussions/5000

**Comment by hifabienne** [comment]
@livio-a @muhlemmer can you maybe help here?

**Comment by muhlemmer** [comment]
The configuration of your app is irrelevant for the OIDC endpoints. OIDC allows all origins as [defined here](https://github.com/zitadel/oidc/blob/6d2092802811f0a9d38cc0a070f2fff708a44181/pkg/op/op.go#L74-L96):

```go
	defaultCORSOptions = cors.Options{
		AllowCredentials: true,
		AllowedHeaders: []string{
			"Origin",
			"Accept",
			"Accept-Language",
			"Authorization",
			"Content-Type",
			"X-Requested-With",
		},
		AllowedMethods: []string{
			http.MethodGet,
			http.MethodHead,
			http.MethodPost,
		},
		ExposedHeaders: []string{
			"Location",
			"Content-Length",
		},
		AllowOriginFunc: func(_ string) bool {
			return true
		},
	}
```

When I send a cURL to my ZITADEL cloud I get the correct headers:

```
curl -H "Origin: http://example.com" --verbose  "https://tims-zitadel-instance-oj7iry.zitadel.cloud/oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code"
*   Trying [2600:1901:0:fa85::]:443...
* Connected to tims-zitadel-instance-oj7iry.zitadel.cloud (2600:1901:0:fa85::) port 443 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.zitadel.cloud
*  start date: Nov 13 23:02:25 2024 GMT
*  expire date: Feb 11 23:58:20 2025 GMT
*  subjectAltName: host "tims-zitadel-instance-oj7iry.zitadel.cloud" matched cert's "*.zitadel.cloud"
*  issuer: C=US; O=Google Trust Services; CN=WR3
*  SSL certificate verify ok.
* using HTTP/2
* h2h3 [:method: GET]
* h2h3 [:path: /oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code]
* h2h3 [:scheme: https]
* h2h3 [:authority: tims-zitadel-instance-oj7iry.zitadel.cloud]
* h2h3 [user-agent: curl/7.88.1]
* h2h3 [accept: */*]
* h2h3 [origin: http://example.com]
* Using Stream ID: 1 (easy handle 0x55a05b873ce0)
> GET /oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code HTTP/2
> Host: tims-zitadel-instance-oj7iry.zitadel.cloud
> user-agent: curl/7.88.1
> accept: */*
> origin: http://example.com
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/2 400 
< access-control-allow-credentials: true
< access-control-allow-origin: http://example.com
< access-control-expose-headers: Location, Content-Length
< cache-control: no-store
< content-type: application/json
< expires: Tue, 26 Nov 2024 07:37:50 GMT
< pragma: no-cache
< set-cookie: __Host-zitadel.useragent=MTczMjYxMDI3MHxFNWRPU0NmZFUwU01nUkJadW9ublBtRnRvdFlhQWlhRXJTcGNCWUNmc1VwZEF5SnR0REtqSFdnbFpUN2VneHBSWVhyVWsySVhKVThHRk1maFRTdGlnYXF0dHRtUUN3PT18OblYGhuNtkn3dQFg7O-OL9G94tPcmKBSYmRy46q7Meo=; Path=/; Max-Age=31536000; HttpOnly; Secure; SameSite=Lax
< set-cookie: zitadel.quota.exhausted=; Path=/; Max-Age=0; SameSite=Lax
< vary: Origin,Cookie
< x-robots-tag: none
< content-length: 70
< date: Tue, 26 Nov 2024 08:37:50 GMT
< server: Google Frontend
< traceparent: 00-96e63c601518507f85ee202d10b25f27-0dab7f74e6e2f51d-00
< x-cloud-trace-context: 96e63c601518507f85ee202d10b25f27/985021083577152797
< via: 1.1 google
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< x-cache-hit: miss
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
<
```

**Reply by muhlemmer** [reply]
Same when I do a preflight btw:

```
curl -X OPTIONS -H "access-control-request-method: GET" -H "Origin: http://example.com" --verbose  "https://tims-zitadel-instance-oj7iry.zitadel.cloud/oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code"
*   Trying [2600:1901:0:fa85::]:443...
* Connected to tims-zitadel-instance-oj7iry.zitadel.cloud (2600:1901:0:fa85::) port 443 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.zitadel.cloud
*  start date: Nov 13 23:02:25 2024 GMT
*  expire date: Feb 11 23:58:20 2025 GMT
*  subjectAltName: host "tims-zitadel-instance-oj7iry.zitadel.cloud" matched cert's "*.zitadel.cloud"
*  issuer: C=US; O=Google Trust Services; CN=WR3
*  SSL certificate verify ok.
* using HTTP/2
* h2h3 [:method: OPTIONS]
* h2h3 [:path: /oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code]
* h2h3 [:scheme: https]
* h2h3 [:authority: tims-zitadel-instance-oj7iry.zitadel.cloud]
* h2h3 [user-agent: curl/7.88.1]
* h2h3 [accept: */*]
* h2h3 [access-control-request-method: GET]
* h2h3 [origin: http://example.com]
* Using Stream ID: 1 (easy handle 0x559c5f1c4ce0)
> OPTIONS /oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code HTTP/2
> Host: tims-zitadel-instance-oj7iry.zitadel.cloud
> user-agent: curl/7.88.1
> accept: */*
> access-control-request-method: GET
> origin: http://example.com
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/2 204 
< access-control-allow-credentials: true
< access-control-allow-methods: GET
< access-control-allow-origin: http://example.com
< vary: Origin, Access-Control-Request-Method, Access-Control-Request-Headers
< x-robots-tag: none
< date: Tue, 26 Nov 2024 08:52:42 GMT
< server: Google Frontend
< traceparent: 00-2ed0d821c663009016720d60b58559c3-02e5d7338f78168d-00
< x-cloud-trace-context: 2ed0d821c663009016720d60b58559c3/208809574192912013
< via: 1.1 google
< strict-transport-security: max-age=63072000; includeSubDomains; preload
< x-cache-hit: uncacheable
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
< 
```

**Reply by guigui42** [reply]
> The configuration of your app is irrelevant for the OIDC endpoints. OIDC allows all origins as [defined here](https://github.com/zitadel/oidc/blob/6d2092802811f0a9d38cc0a070f2fff708a44181/pkg/op/op.go#L74-L96):
> 
> ```go
> 	defaultCORSOptions = cors.Options{
> 		AllowCredentials: true,
> 		AllowedHeaders: []string{
> 			"Origin",
> 			"Accept",
> 			"Accept-Language",
> 			"Authorization",
> 			"Content-Type",
> 			"X-Requested-With",
> 		},
> 		AllowedMethods: []string{
> 			http.MethodGet,
> 			http.MethodHead,
> 			http.MethodPost,
> 		},
> 		ExposedHeaders: []string{
> 			"Location",
> 			"Content-Length",
> 		},
> 		AllowOriginFunc: func(_ string) bool {
> 			return true
> 		},
> 	}
> ```
> 
> When I send a cURL to my ZITADEL cloud I get the correct headers:
> 
> ```
> curl -H "Origin: http://example.com" --verbose  "https://tims-zitadel-instance-oj7iry.zitadel.cloud/oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code"
> *   Trying [2600:1901:0:fa85::]:443...
> * Connected to tims-zitadel-instance-oj7iry.zitadel.cloud (2600:1901:0:fa85::) port 443 (#0)
> * ALPN: offers h2,http/1.1
> * TLSv1.3 (OUT), TLS handshake, Client hello (1):
> *  CAfile: /etc/ssl/certs/ca-certificates.crt
> *  CApath: /etc/ssl/certs
> * TLSv1.3 (IN), TLS handshake, Server hello (2):
> * TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
> * TLSv1.3 (IN), TLS handshake, Certificate (11):
> * TLSv1.3 (IN), TLS handshake, CERT verify (15):
> * TLSv1.3 (IN), TLS handshake, Finished (20):
> * TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
> * TLSv1.3 (OUT), TLS handshake, Finished (20):
> * SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
> * ALPN: server accepted h2
> * Server certificate:
> *  subject: CN=*.zitadel.cloud
> *  start date: Nov 13 23:02:25 2024 GMT
> *  expire date: Feb 11 23:58:20 2025 GMT
> *  subjectAltName: host "tims-zitadel-instance-oj7iry.zitadel.cloud" matched cert's "*.zitadel.cloud"
> *  issuer: C=US; O=Google Trust Services; CN=WR3
> *  SSL certificate verify ok.
> * using HTTP/2
> * h2h3 [:method: GET]
> * h2h3 [:path: /oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code]
> * h2h3 [:scheme: https]
> * h2h3 [:authority: tims-zitadel-instance-oj7iry.zitadel.cloud]
> * h2h3 [user-agent: curl/7.88.1]
> * h2h3 [accept: */*]
> * h2h3 [origin: http://example.com]
> * Using Stream ID: 1 (easy handle 0x55a05b873ce0)
> > GET /oauth/v2/authorize?client_id=294646046196301826&redirect_uri=http%3A%2F%2Flocalhost%3A9999%2Fauth%2Fcallback&scope=oidc&response_type=code HTTP/2
> > Host: tims-zitadel-instance-oj7iry.zitadel.cloud
> > user-agent: curl/7.88.1
> > accept: */*
> > origin: http://example.com
> > 
> * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
> * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
> * old SSL session ID is stale, removing
> < HTTP/2 400 
> < access-control-allow-credentials: true
> < access-control-allow-origin: http://example.com
> < access-control-expose-headers: Location, Content-Length
> < cache-control: no-store
> < content-type: application/json
> < expires: Tue, 26 Nov 2024 07:37:50 GMT
> < pragma: no-cache
> < set-cookie: __Host-zitadel.useragent=MTczMjYxMDI3MHxFNWRPU0NmZFUwU01nUkJadW9ublBtRnRvdFlhQWlhRXJTcGNCWUNmc1VwZEF5SnR0REtqSFdnbFpUN2VneHBSWVhyVWsySVhKVThHRk1maFRTdGlnYXF0dHRtUUN3PT18OblYGhuNtkn3dQFg7O-OL9G94tPcmKBSYmRy46q7Meo=; Path=/; Max-Age=31536000; HttpOnly; Secure; SameSite=Lax
> < set-cookie: zitadel.quota.exhausted=; Path=/; Max-Age=0; SameSite=Lax
> < vary: Origin,Cookie
> < x-robots-tag: none
> < content-length: 70
> < date: Tue, 26 Nov 2024 08:37:50 GMT
> < server: Google Frontend
> < traceparent: 00-96e63c601518507f85ee202d10b25f27-0dab7f74e6e2f51d-00
> < x-cloud-trace-context: 96e63c601518507f85ee202d10b25f27/985021083577152797
> < via: 1.1 google
> < strict-transport-security: max-age=63072000; includeSubDomains; preload
> < x-cache-hit: miss
> < alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
> <
> ```

Thanks @muhlemmer 
Also I have checked and it also happens locally, so it is not related to fly.io
Could it be because I am using HTMX, using some more specific headers like hx-current-url,hx-request,hx-trigger-name,x-workspace-id ?

**Reply by muhlemmer** [reply]
Could very well be, but I don't know the specifics of the CORS library like that. Just try without those extra headers, or use curl with and without those extra headers and you will know.

**Comment by guigui42** [comment]
@muhlemmer ok I can confirm the issue is with the HTMX headers
Standard curl query like the one you gave, with my domain and own zitadel domain
I get 

```
< HTTP/2 204 
< access-control-allow-credentials: true
< access-control-allow-methods: GET
< access-control-allow-origin: https://xxxxxx
< vary: Origin, Access-Control-Request-Method, Access-Control-Request-Headers
```

But as soon as I put back those HTMX headers, it fails
` -H 'access-control-request-headers: hx-current-url,hx-request,hx-trigger-name' `

```
< HTTP/2 204 
< vary: Origin, Access-Control-Request-Method, Access-Control-Request-Headers

```

I see 4 ways of fixing it (tell me if am wrong) : 

1. Zitadel adding extra allowed headers server side
2. zitadel-go/v3/pkg/authentication getting updated to allow to strip these headers from the [middleware](https://pkg.go.dev/github.com/zitadel/zitadel-go/v3@v3.3.1/pkg/authentication#Interceptor.RequireAuthentication) when communicating with Zitadel
3. rewriting the zitadel-go go middleware for this use case
4. I remove those headers from my front-end using hx-request='{"noHeaders": true}', but I lose some functionalities (see [documented issue](https://github.com/bigskysoftware/htmx/issues/779))

Any advice on how to proceed ?

**Comment by guigui42** [comment, marked as answer]
Alright, if anyone is interested, I made a wrapper for the Zitadel middleware to strip unecessary headers and putting them back in after :

```go
type ZitadelAuthMiddleware = authentication.Interceptor[*openid.UserInfoContext[*oidc.IDTokenClaims, *oidc.UserInfo]]

type responseWriter struct {
	http.ResponseWriter
	status      int
	wroteHeader bool
}

func (rw *responseWriter) WriteHeader(code int) {
	if !rw.wroteHeader {
		rw.status = code
		rw.ResponseWriter.WriteHeader(code)
		rw.wroteHeader = true
	}
}

func (rw *responseWriter) Status() int {
	return rw.status
}


func (m *AuthMiddleware) WrapZitadelAuth(mw *ZitadelAuthMiddleware) func(next http.Handler) http.Handler {
	auth := mw.RequireAuthentication()
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Save original headers
			originalHeaders := make(map[string][]string)
			if r.Header.Get("HX-Request") == "true" {
				htmxPrefixes := []string{"Cf-", "Fly-", "HX-", "X-"}

				// Save and remove HTMX-related headers
				for headerName, headerValues := range r.Header {
					for _, prefix := range htmxPrefixes {
						if strings.HasPrefix(strings.ToUpper(headerName), strings.ToUpper(prefix)) {
							originalHeaders[headerName] = headerValues
							r.Header.Del(headerName)
						}
					}
				}
			}

			// Run auth with cleaned headers
			wrapped := &responseWriter{ResponseWriter: w}

			// Create handler that will restore headers after auth
			authNext := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				// Restore original headers
				for name, values := range originalHeaders {
					r.Header[name] = values
				}
				next.ServeHTTP(w, r)
			})

			auth(authNext).ServeHTTP(wrapped, r)
		})
	}
}
```

Then you can just call it this way (in this case I use Chi) :

```go
// Zitadel-go sdk
mw := authentication.Middleware(authN)
r.Use(authMiddleware.WrapZitadelAuth(mw))
```
